WordPress/gutenberg
# Block variations: Add block-supports flag to add variation specific classname

Related to: https://github.com/WordPress/gutenberg/issues/43743

## What?
For a given variation of a static block that is registered it will add a unique className specific to the variation to the block markup. For example [`group-row`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/group/variations.js#L22-L33) which is a variant of `core/group` would have the resulting CSS className: `wp-block-group wp-block-group-group-row`.

## Why?
As part of the project to support block aliases for variations ([tracking issue](https://github.com/WordPress/gutenberg/issues/43743)) one of the requirements is to provide an additional CSS className to provide more styling options for these blocks.

## How?

We chose to update the existing `className` block-support flag. Previously, setting `className: true` would cause the default block classname (`wp-block-my-block`) to be added. This continues to be the default value and behavior for that flag.

Additionally, the following syntax is now supported:
```js
className: {
    block: true,
    variation: true,
}
```

The `block` property controls adding the default block classname, whereas the `variation` property is in charge of adding the variation-specific classname (`wp-block-my-variation`).

**Question:** Are we happy with the choice of names for these properties (`block`, `variation`)? Should they be called differently?

## Testing Instructions

* Apply the patch below, which will opt the Group block into variation classname block support.
* Using the TT4 theme, go to the Site Editor.
* In the DevTools console, verify that deprecations have been run for all instances of the Group block, and that they've injected the new variation-specific classnames. (See screenshot at the bottom of the PR desc.)
* In the Code Editor, verify that all Group blocks have a classname added for their respective variation (e.g. `wp-block-group-group-row`).
* Save the template (to persist changes made by block deprecations).
* Reload. Verify that no blocks show up as invalid.
* View the frontend.
* In the developer tools, verify that the classname is present on the frontend as well.
* Go to the post editor and insert a few new Group blocks. Use different variations, and verify in the Code Editor that each of them gets the correct variation classname added.

<details>
<summary>
Patch
</summary>

```diff
diff --git a/packages/block-library/src/group/block.json b/packages/block-library/src/group/block.json
index 3c7d8d3ce13..bb958b0cce6 100644
--- a/packages/block-library/src/group/block.json
+++ b/packages/block-library/src/group/block.json
@@ -91,6 +91,9 @@
 		},
 		"interactivity": {
 			"clientNavigation": true
+		},
+		"className": {
+			"variation": true
 		}
 	},
 	"editorStyle": "wp-block-group-editor",
diff --git a/packages/block-library/src/group/deprecated.js b/packages/block-library/src/group/deprecated.js
index fcc2b249d22..b97af894eeb 100644
--- a/packages/block-library/src/group/deprecated.js
+++ b/packages/block-library/src/group/deprecated.js
@@ -42,6 +42,100 @@ const migrateAttributes = ( attributes ) => {
 };
 
 const deprecated = [
+	// Version without block support 'className: { variation: true }'.
+	{
+		attributes: {
+			tagName: {
+				type: 'string',
+				default: 'div',
+			},
+			templateLock: {
+				type: [ 'string', 'boolean' ],
+				enum: [ 'all', 'insert', 'contentOnly', false ],
+			},
+			allowedBlocks: {
+				type: 'array',
+			},
+		},
+		supports: {
+			__experimentalOnEnter: true,
+			__experimentalOnMerge: true,
+			__experimentalSettings: true,
+			align: [ 'wide', 'full' ],
+			anchor: true,
+			ariaLabel: true,
+			html: false,
+			background: {
+				backgroundImage: true,
+				backgroundSize: true,
+				__experimentalDefaultControls: {
+					backgroundImage: true,
+				},
+			},
+			color: {
+				gradients: true,
+				heading: true,
+				button: true,
+				link: true,
+				__experimentalDefaultControls: {
+					background: true,
+					text: true,
+				},
+			},
+			shadow: true,
+			spacing: {
+				margin: [ 'top', 'bottom' ],
+				padding: true,
+				blockGap: true,
+				__experimentalDefaultControls: {
+					padding: true,
+					blockGap: true,
+				},
+			},
+			dimensions: {
+				minHeight: true,
+			},
+			__experimentalBorder: {
+				color: true,
+				radius: true,
+				style: true,
+				width: true,
+				__experimentalDefaultControls: {
+					color: true,
+					radius: true,
+					style: true,
+					width: true,
+				},
+			},
+			position: {
+				sticky: true,
+			},
+			typography: {
+				fontSize: true,
+				lineHeight: true,
+				__experimentalFontFamily: true,
+				__experimentalFontWeight: true,
+				__experimentalFontStyle: true,
+				__experimentalTextTransform: true,
+				__experimentalTextDecoration: true,
+				__experimentalLetterSpacing: true,
+				__experimentalDefaultControls: {
+					fontSize: true,
+				},
+			},
+			layout: {
+				allowSizingOnChildren: true,
+			},
+			interactivity: {
+				clientNavigation: true,
+			},
+		},
+		save( { attributes: { tagName: Tag } } ) {
+			return (
+				<Tag { ...useInnerBlocksProps.save( useBlockProps.save() ) } />
+			);
+		},
+	},
 	// Version with default layout.
 	{
 		attributes: {
```
</details>

## Screenshot

<img width="519" alt="image" src="https://github.com/user-attachments/assets/2af6d479-4b8d-43a7-a6a8-3d30453479a4">
